added xml docs to concepts (#115)
IRL2/nanover-docs: source/tutorials/openmm.rst

OpenMM
======

.. contents:: Contents
    :depth: 2
    :local:

> Openmm Tutorials

Reviewer: I've just noticed that "OpenMM" is written "Openmm" in a few places on this page. We should check and make sure that we consistently write "OpenMM".
```suggestion
OpenMM Tutorials
```

  NanoVer simulation using a python client.
* `openmm_neuraminidase`: A notebook that demonstrates how to set up an interactive
  OpenMM simulation from scratch with NanoVer and change the visualisations of
  different atom selections for a protein-ligand system.

> XML Files

Reviewer: I think maybe this heading ("XML Files") could be refactored to something that makes it clear that these are NanoVer simulation files of an OpenMM simulation, rather than being a general XML file. It could also be good to add a sentence below this heading to explain this.

Reviewer 2: Perhaps something like "NanoVer OpenMM input files"? XML is just a general nested file structure, not specific to NanoVer or OpenMM, and the important thing is that the input files described in this section are specific to OpenMM simulations performed using NanoVer

Author: The XML format we are using does not seem to be different from the one OpenMM defines in their documentation. As far as I know, it is not proprietary to NanoVer (The same Serializer Class exists in Openmm with the same XML structure [here](http://docs.openmm.org/latest/api-python/generated/openmm.openmm.XmlSerializer.html?highlight=serializer#openmm.openmm.XmlSerializer) )

Reviewer: So there is a slight difference between the OpenMM format and the NanoVer format. NanoVer input files for OpenMM simulations contain a pdb and a number of xml serialized objects. Although the serialized objects are identical to those used by OpenMM (they are actually created used OpenMM), the overall contents and structure of the files is specific to NanoVer and cannot be read by OpenMM on it's own. Explaining this could prevent users misunderstanding the NanoVer input file format.

Author: Thanks for clearing this out @rhoslynroebuck. I agree that specifying the overall contents and structure of the XML files in the docs would prevent users from misinterpreting or misusing them.

Author: Resolved in commit [f62ca6e](https://github.com/IRL2/nanover-docs/pull/115/commits/f62ca6e5fd4d7a15e0d7843f064081e5d054ccf2)

==============================================

The XML format for serializing OpenMM simulations consists of a root element ``<OpenMMSimulation>`` containing three main components:

1. **Starting Structure**: Enclosed in either ``<pdbx>`` or ``<pdb>`` tags
2. **OpenMM Serialized System**: Enclosed in the ``<System>`` tag
3. **OpenMM Serialized Integrator**: Enclosed in the ``<Integrator>`` tag

An optional fourth component, the OpenMM serialized state, may also be included.

> The format allows for efficient storage and exchange of OpenMM simulation data, including the starting structure, system configuration, and integrator settings.

Reviewer: Can you link to some documentation about what an "OpenMM serialized state" is?

Author: I read about it in the openmm documentation [here](http://docs.openmm.org/7.7.0/userguide/theory/05_other_features.html?highlight=serialized%20state#xmlserializer) but nothing more exists about it. I wanted to keep the information there in case someone know what it is and needs it.

Reviewer 2: I think that OpenMM just has a method of serializing the objects it uses (be is a system, integrator, state etc.), and the label "OpenMM serialized" just indicates that it's been created using the functionality of OpenMM. So rather than explaining what an OpenMM serialized state is, it may be better to put in a short sentence to say that the fields prefixed by "OpenMM serialized" are created using the OpenMM inbuilt functionality to do so.

Author: Solved in this commit [0043457](https://github.com/IRL2/nanover-docs/pull/115/commits/0043457c7c21028f743a19611dc6a729dd6a82c8)

> XML Structure

Reviewer: Each word starts with an upper case in the sub headings on this page, which is inconsistent with the rest of the docs. Would be good to change these so only the first letter starts with an upper case.

-------------

.. code-block:: xml

    <OpenMMSimulation>
        <pdbx>
            <!-- Content of the PDBx file -->
        </pdbx>
        <System ...>
            <!-- XML content of the OpenMM serialized system -->
        </System>
        <Integrator ...>
            <!-- XML content of the OpenMM serialized integrator -->
        </Integrator>

>     </OpenMMSimulation>

Reviewer 2: May be worth keeping an eye on IRL2/nanover-protocol#285, as this will add the `State` field too

Author: Done in this commit [0043457](https://github.com/IRL2/nanover-docs/pull/115/commits/0043457c7c21028f743a19611dc6a729dd6a82c8).

Components Description
----------------------

1. **Starting Structure**:
^^^^^^^^^^^^^^^^^^^^^^^^^^

   * Tag: ``<pdbx>`` (preferred) or ``<pdb>`` (for backward compatibility)
   * Content: The entire PDBx or PDB file content

2. **OpenMM Serialized System**:
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Within this tag, you'll find:

>     **Particles:** They are represented with ``<Particle>`` tags, each containing:

Reviewer: This line is all in bold (default for a nested list in Sphinx). Adding an extra line break stops this (https://stackoverflow.com/questions/42297839/parent-element-of-nested-list-is-bold)

        .. code-block:: xml

           <Constraints>
             <Constraint p1="0" p2="1" distance="0.1"/>
           </Constraints>

3. **OpenMM Serialized Integrator**:
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The ``<Integrator>`` tag contains parameters that specify the integration method to be used to simulate dynamics, such as the type of integrator, simulation time step and temperature:

.. code-block:: xml

    <Integrator type="LangevinIntegrator" constraintTolerance="1e-05" friction="4" randomSeed="0" stepSize=".0005" temperature="300" version="1" />

More details on integrators can be found `here <http://docs.openmm.org/latest/userguide/theory/04_integrators.html>`_.

Usage
-----

The :mod:`nanover.openmm.serializer` module provides the ``serialize_simulation`` and ``deserialize_simulation`` functions which allow saving and loading OpenMM simulations to/from XML files. The serialization captures the complete simulation state including:

- Structure coordinates and topology (as PDBx/PDB)
- OpenMM System definition
- Integrator configuration

Serializing a Simulation
^^^^^^^^^^^^^^^^^^^^^^^^

> To save a simulation to XML::

Reviewer: Can you make this into a code block?

    xml_string = nanover.openmm.serializer.serialize_simulation(simulation)

    with open("sim.xml", "w") as f:
        f.write(xml_string)

Deserializing a Simulation
^^^^^^^^^^^^^^^^^^^^^^^^^^

> To load a simulation from XML::

Reviewer: I think it would be good to make it clear that this is loading a simulation from a NanoVer OpenMM XML file, not just a general XML file. (Same comment for below)

Author: resolved in commit [f62ca6e](https://github.com/IRL2/nanover-docs/pull/115/commits/f62ca6e5fd4d7a15e0d7843f064081e5d054ccf2)

>     with open("sim.xml", "r") as f:

Reviewer: Can you make this into a code block?

        simulation = nanover.openmm.serializer.deserialize_simulation(f.read())

The ``deserialize_simulation`` function accepts optional arguments:

- ``imd_force``: A CustomExternalForce for interactive molecular dynamics
- ``platform_name``: The OpenMM platform to use (e.g. "CUDA", "OpenCL")

For example::

>     simulation = nanover.openmm.serializer.deserialize_simulation(

Reviewer: Can you make this into a code block?

  NanoVer simulation using a python client.
* `openmm_neuraminidase`: A notebook that demonstrates how to set up an interactive
  OpenMM simulation from scratch with NanoVer and change the visualisations of
  different atom selections for a protein-ligand system.

Openmm XML Files
==============================================

> The XML format described here is used specifically for manipulating NanoVer simulations from OpenMM serialized XML files. It consists of a root element ``<OpenMMSimulation>`` containing three main components:

Reviewer 2: ```suggestion
The XML format described here is used specifically for saving and loading NanoVer OpenMM simulations. It consists of a root element ``<OpenMMSimulation>`` containing three main components:
```
The suggested change here reflects the fact that the xml file also contains non-OpenMM components from the PDB file.

Usage
-----

The :mod:`nanover.openmm.serializer` module provides the ``serialize_simulation`` and ``deserialize_simulation`` functions which allow saving and loading OpenMM simulations to/from XML files. The serialization captures the complete simulation state including:

- Structure coordinates and topology (as PDBx/PDB)
- OpenMM System definition
- Integrator configuration

Serializing a simulation
^^^^^^^^^^^^^^^^^^^^^^^^

To save a simulation to an Openmm XML::

    xml_string = nanover.openmm.serializer.serialize_simulation(simulation)

    with open("sim.xml", "w") as f:
        f.write(xml_string)

Deserializing a simulation
^^^^^^^^^^^^^^^^^^^^^^^^^^

To load a simulation from an Openmm XML::

    with open("sim.xml", "r") as f:
        simulation = nanover.openmm.serializer.deserialize_simulation(f.read())

The ``deserialize_simulation`` function accepts optional arguments:

- ``imd_force``: A CustomExternalForce for interactive molecular dynamics
- ``platform_name``: The OpenMM platform to use (e.g. "CUDA", "OpenCL")

For example::

    simulation = nanover.openmm.serializer.deserialize_simulation(
        xml_string,
        platform_name="CUDA"
    )

> 

Reviewer 2: May be worth linking to an OpenMM tutorial that demonstrates how to write OpenMM input files (e.g. the polyalanine notebook in the OpenMM examples directory)

Author: Is [this](https://github.com/openmm/openmm-cookbook/blob/main/notebooks/cookbook/Saving%20Systems%20to%20XML%20Files.ipynb) what you mean?

Reviewer: I think Harry means the NanoVer OpenMM tutorial about polyanalanine: https://github.com/IRL2/nanover-protocol/blob/main/examples/openmm/openmm_polyalanine.ipynb

Author: resolved in commit [e43e5bf](https://github.com/IRL2/nanover-docs/pull/115/commits/e43e5bf8a8b24a180e5f9e6fa04ecfc6947ae248)
